File: frontend/src/api/wallet.ts
/* [154A-15a] API client para wallet y solicitudes de cancelación.
 * GET /api/wallet — saldo actual
 * GET /api/wallet/transactions — historial paginado
 * GET /api/orders/:id/cancel-request — listar solicitudes de cancelación
 * POST /api/orders/:id/cancel-request — crear solicitud de cancelación
 * POST /api/orders/:id/cancel-request/:reqId/respond — aceptar/rechazar */
import instance from './axios-instance';

export interface WalletResponse {
    id: string;
    user_id: string;
    balance_cents: number;
    currency: string;
}

export interface WalletTransactionResponse {
    id: string;
    wallet_id: string;
    user_id: string;
    amount_cents: number;
    transaction_type: string;
    reference_type: string | null;
    reference_id: string | null;
    description: string | null;
    balance_after_cents: number;
    created_at: string;
}

export interface WalletTransactionsPage {
    items: WalletTransactionResponse[];
    total: number;
    page: number;
    per_page: number;
}

export interface CancellationRequestResponse {
    id: string;
    order_id: string;
    order_number: number | null;
    requested_by: string;
    requester_name: string | null;
    reason: string;
    status: string;
    resolved_by: string | null;
    resolved_at: string | null;
    created_at: string;
}

/* Helpers para formatear saldo */
export function formatBalance(cents: number, currency = 'USD'): string {
    return new Intl.NumberFormat('es', {
        style: 'currency',
        currency,
        minimumFractionDigits: 2,
    }).format(cents / 100);
}

/* GET /api/wallet */
export async function apiGetBalance(): Promise<WalletResponse> {
    const { data } = await instance.get<WalletResponse>('/api/wallet');
    return data;
}

/* GET /api/wallet/transactions */
export async function apiListTransactions(
    page = 1,
    perPage = 20
): Promise<WalletTransactionsPage> {
    const { data } = await instance.get<WalletTransactionsPage>(
        '/api/wallet/transactions',
        { params: { page, per_page: perPage } }
    );
    return data;
}

/* --- Withdrawal (retiro de saldo) --- */

export interface WithdrawalRequestResponse {
    id: string;
    user_id: string;
    amount_cents: number;
    status: string;
    payment_method: string | null;
    payment_details: string | null;
    admin_notes: string | null;
    resolved_by: string | null;
    created_at: string;
    resolved_at: string | null;
}

export interface WithdrawalRequestsPage {
    items: WithdrawalRequestResponse[];
    total: number;
    page: number;
    per_page: number;
}

export interface CreateWithdrawalRequest {
    amount_cents: number;
    payment_method?: string;
    payment_details?: string;
}

export interface ResolveWithdrawalRequest {
    approve: boolean;
    admin_notes?: string;
}

/* POST /api/wallet/withdraw */
export async function apiCreateWithdrawal(body: CreateWithdrawalRequest): Promise<WithdrawalRequestResponse> {
    const { data } = await instance.post<WithdrawalRequestResponse>('/api/wallet/withdraw', body);
    return data;
}

/* GET /api/wallet/withdrawals */
export async function apiListWithdrawals(page = 1, perPage = 10): Promise<WithdrawalRequestsPage> {
    const { data } = await instance.get<WithdrawalRequestsPage>(
        '/api/wallet/withdrawals',
        { params: { page, per_page: perPage } }
    );
    return data;
}

/* GET /api/admin/withdrawals */
export async function apiAdminListWithdrawals(page = 1, perPage = 20): Promise<WithdrawalRequestsPage> {
    const { data } = await instance.get<WithdrawalRequestsPage>(
        '/api/admin/withdrawals',
        { params: { page, per_page: perPage } }
    );
    return data;
}

/* PATCH /api/admin/withdrawals/:id */
export async function apiResolveWithdrawal(id: string, body: ResolveWithdrawalRequest): Promise<WithdrawalRequestResponse> {
    const { data } = await instance.patch<WithdrawalRequestResponse>(`/api/admin/withdrawals/${id}`, body);
    return data;
}

/* POST /api/orders/:orderId/cancel-request */
export async function apiCreateCancellationRequest(
    orderId: string,
    reason: string
): Promise<CancellationRequestResponse> {
    const { data } = await instance.post<CancellationRequestResponse>(
        `/api/orders/${orderId}/cancel-request`,
        { reason }
    );
    return data;
}

/* GET /api/orders/:orderId/cancel-request — solicitudes pendientes */
export async function apiGetCancellationRequests(
    orderId: string
): Promise<CancellationRequestResponse[]> {
    const { data } = await instance.get<CancellationRequestResponse[]>(
        `/api/orders/${orderId}/cancel-request`
    );
    return data;
}

/* POST /api/orders/:orderId/cancel-request/:requestId/respond */
export async function apiRespondCancellationRequest(
    orderId: string,
    requestId: string,
    accept: boolean
): Promise<CancellationRequestResponse> {
    const { data } = await instance.post<CancellationRequestResponse>(
        `/api/orders/${orderId}/cancel-request/${requestId}/respond`,
        { accept }
    );
    return data;
}

File: src/models/wallet.rs
/* [154A-15a] Wallet system: modelos para saldo virtual de usuarios.
 * UserWallet: saldo en cents por usuario (uno a uno).
 * WalletTransaction: cada movimiento de saldo con snapshot del balance.
 * CancellationRequest: solicitud de cancelación con flujo de aprobación.
 * [184A-1] WithdrawalRequest: solicitudes de retiro de fondos del wallet. */

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use utoipa::ToSchema;
use uuid::Uuid;

#[derive(Debug, Clone, FromRow)]
pub struct UserWallet {
    pub id: Uuid,
    pub user_id: Uuid,
    pub balance_cents: i32,
    pub currency: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow)]
pub struct WalletTransaction {
    pub id: Uuid,
    pub wallet_id: Uuid,
    pub user_id: Uuid,
    pub amount_cents: i32,
    pub transaction_type: String,
    pub reference_type: Option<String>,
    pub reference_id: Option<Uuid>,
    pub description: Option<String>,
    pub balance_after_cents: i32,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow)]
pub struct CancellationRequest {
    pub id: Uuid,
    pub order_id: Uuid,
    pub requested_by: Uuid,
    pub reason: String,
    pub status: String,
    pub resolved_by: Option<Uuid>,
    pub resolved_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
}

/* ============================================================
   RESPONSES
   ============================================================ */

/* [184A-1] Withdrawal request: solicitud de retiro de fondos */
#[derive(Debug, Clone, FromRow)]
pub struct WithdrawalRequest {
    pub id: Uuid,
    pub user_id: Uuid,
    pub amount_cents: i32,
    pub status: String,
    pub payment_method: Option<String>,
    pub payment_details: Option<String>,
    pub admin_notes: Option<String>,
    pub resolved_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub resolved_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct WithdrawalRequestResponse {
    pub id: Uuid,
    pub user_id: Uuid,
    pub amount_cents: i32,
    pub status: String,
    pub payment_method: Option<String>,
    pub payment_details: Option<String>,
    pub admin_notes: Option<String>,
    pub resolved_by: Option<Uuid>,
    pub created_at: String,
    pub resolved_at: Option<String>,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct WithdrawalRequestsPage {
    pub requests: Vec<WithdrawalRequestResponse>,
    pub total: i64,
    pub page: i64,
    pub per_page: i64,
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct CreateWithdrawalRequest {
    pub amount_cents: i32,
    pub payment_method: Option<String>,
    pub payment_details: Option<String>,
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct ResolveWithdrawalRequest {
    pub approve: bool,
    pub admin_notes: Option<String>,
}

impl From<&WithdrawalRequest> for WithdrawalRequestResponse {
    fn from(w: &WithdrawalRequest) -> Self {
        Self {
            id: w.id,
            user_id: w.user_id,
            amount_cents: w.amount_cents,
            status: w.status.clone(),
            payment_method: w.payment_method.clone(),
            payment_details: w.payment_details.clone(),
            admin_notes: w.admin_notes.clone(),
            resolved_by: w.resolved_by,
            created_at: w.created_at.to_rfc3339(),
            resolved_at: w.resolved_at.map(|t| t.to_rfc3339()),
        }
    }
}

#[derive(Debug, Serialize, ToSchema)]
pub struct WalletResponse {
    pub id: Uuid,
    pub user_id: Uuid,
    pub balance_cents: i32,
    pub currency: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct WalletTransactionResponse {
    pub id: Uuid,
    pub amount_cents: i32,
    pub transaction_type: String,
    pub reference_type: Option<String>,
    pub reference_id: Option<Uuid>,
    pub description: Option<String>,
    pub balance_after_cents: i32,
    pub created_at: String,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct WalletTransactionsPage {
    pub transactions: Vec<WalletTransactionResponse>,
    pub total: i64,
    pub page: i64,
    pub per_page: i64,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct CancellationRequestResponse {
    pub id: Uuid,
    pub order_id: Uuid,
    pub order_number: Option<i32>,
    pub requested_by: Uuid,
    pub requester_name: Option<String>,
    pub reason: String,
    pub status: String,
    pub resolved_by: Option<Uuid>,
    pub resolved_at: Option<String>,
    pub created_at: String,
}

/* ============================================================
   REQUESTS
   ============================================================ */

#[derive(Debug, Deserialize, ToSchema)]
pub struct CreateCancellationRequest {
    pub reason: String,
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct RespondCancellationRequest {
    pub accept: bool,
}

/* ============================================================
   CONVERSIONS
   ============================================================ */

impl From<&UserWallet> for WalletResponse {
    fn from(w: &UserWallet) -> Self {
        Self {
            id: w.id,
            user_id: w.user_id,
            balance_cents: w.balance_cents,
            currency: w.currency.clone(),
            created_at: w.created_at.to_rfc3339(),
            updated_at: w.updated_at.to_rfc3339(),
        }
    }
}

impl From<&WalletTransaction> for WalletTransactionResponse {
    fn from(t: &WalletTransaction) -> Self {
        Self {
            id: t.id,
            amount_cents: t.amount_cents,
            transaction_type: t.transaction_type.clone(),
            reference_type: t.reference_type.clone(),
            reference_id: t.reference_id,
            description: t.description.clone(),
            balance_after_cents: t.balance_after_cents,
            created_at: t.created_at.to_rfc3339(),
        }
    }
}
